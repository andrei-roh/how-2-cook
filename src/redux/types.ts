export const SET_USER = 'SET_USER';
export const SET_LOADING = 'SET_LOADING';
export const SET_ERROR = 'SET_ERROR';
export const ADD_RECIPES_TO_LIST = 'ADD_RECIPES_TO_LIST';
export const UPDATE_RECIPES_LIST = 'UPDATE_RECIPES_LIST';
export const ADD_IMAGES_TO_LIST = 'ADD_IMAGES_TO_LIST';
export const SET_RECIPE_SEARCH_INPUT = 'SET_RECIPE_SEARCH_INPUT';
export const SET_MANAGE_RECIPES_SCROLL_SIZE = 'SET_MANAGE_RECIPES_SCROLL_SIZE';
export const SET_MANAGE_RECIPES_SCROLL_DIRECTION =
  'SET_MANAGE_RECIPES_SCROLL_DIRECTION';
export const SET_SHOW_MENU = 'SET_SHOW_MENU';
export const ADD_INGREDIENTS_TO_LIST = 'ADD_INGREDIENTS_TO_LIST';
export const SET_INGREDIENT_SEARCH_INPUT = 'SET_INGREDIENT_SEARCH_INPUT';
export const UPDATE_INGREDIENTS_LIST = 'UPDATE_INGREDIENTS_LIST';


