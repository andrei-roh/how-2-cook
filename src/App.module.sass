@use 'src/style/colors'

.app 
  padding-top: 46px
  background-color: colors.$white
  height: 100vh
  box-sizing: border-box
  display: flex
  flex-direction: column
  align-items: center

