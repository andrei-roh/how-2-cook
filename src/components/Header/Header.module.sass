@use 'src/style/colors'
@import 'src/style/typography'

.headerWrapper 
    width: 100%
    max-width: 768px
    height: 46px
    background-color: colors.$white
    display: grid
    grid-template-columns: 90% auto
    align-items: center
    justify-content: space-between
    padding: 6px 12px
    margin: 0
    box-sizing: border-box
    border-bottom: 1px solid colors.$gray
    position: fixed
    top: 0
    z-index: 2

    .userEmail
        @extend .H5
        text-align: start

    .headerModalButton
        width: 30px

        .headerLogo
            height: 30px

    .modalCancelButton
        color: colors.$purple
        background-color: colors.$white
        border: 1px solid colors.$purple

        &:hover
            color: colors.$white
            background-color: colors.$blue
            border: none

    .modalUserLogoutButton
        background-color: colors.$blue
        color: colors.$white

        &:hover
            background-color: colors.$sky
