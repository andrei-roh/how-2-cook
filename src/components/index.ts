export { Header } from './Header/Header';
export { Button } from './Button/Button';
export { Input } from './Input/Input';
export { Modal } from './Modal/Modal';
export { Notification } from './Notification/Notification';
export { ImageLoader } from './ImageLoader/ImageLoader';
export { TextArea } from './TextArea/TextArea';
export { Select } from './Select/Select';