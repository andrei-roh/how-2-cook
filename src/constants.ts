export const ROOT_ROUTE = '/';
export const RECIPES_ROUTE = '/recipes';
export const CREATE_RECIPE_ROUTE = '/recipe/create';

export const DEFAULT_NOTIFICATION = 'Default Notification Message';
export const NOTIFICATION = (value?: string) => ({
  USER_WRONG_EMAIL: `Пользователь с адресом электронной почты ${value} не существует. Проверьте данные или обратитесь к администратору`,
  USER_WRONG_PASSWORD: `Введен неправильный пароль для пользователя ${value}. Проверьте данные или обратитесь к администратору`,
  USER_TOO_MANY_REQUESTS: `Слишком много запросов для пользователя ${value}. Подождите некоторое время или обратитесь к администратору`,
  SIGN_OUT_SUCCESS: `Пользователь ${value} вышел из приложения`,
  SIGN_OUT_ERROR: `Ошибка при выходе пользователя ${value} из приложения`,
});
