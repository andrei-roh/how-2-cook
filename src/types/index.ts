import { Action } from '@reduxjs/toolkit';

export interface IUser {
  email: string;
}

export enum DishType {
  Main = 'main',
  Soup = 'soup',
  Dessert = 'dessert',
  Salat = 'salat',
  Snack = 'snack',
}

export type RecipeImage = {
  [key: string]: string;
};

export interface IIngredient {
  id: string;
  name: string;
}

export interface IRecipe {
  id: string;
  imageUrl: string;
  name: string;
  type: DishType;
  isVegan: boolean;
  ingredients: string[];
  description: string;
  composition?: string;
  createdAt?: string;
  createdBy?: string;
  updatedAt?: string;
  updatedBy?: string;
}

export interface IPlanningList {
  id: string;
  name: string;
  ingredientsList: string[];
  recipesList: string[];
}

export interface IState {
  user: IUser;
  loading: boolean;
  error: null | string;
  recipesList: IRecipe[];
  imagesList: RecipeImage[];
  recipesPage: {
    recipeSearchInput: string;
    scrollSize: number;
    scrollDirection: ScrollDirection;
  };
  isMenu: boolean;
  ingredientsList: IIngredient[];
  ingredientsPage: {
    ingredientSearchInput: string;
  }
  previousRoute: string;
  planningPage: {
    ingredientsList: string[];
    recipesList: IRecipe[];
  }
}

export interface IAction extends Action {
  type: string;
  payload: never;
}

export interface IValidationError {
  [key: string]: string;
}

export enum FirebaseErrorType {
  InvalidEmail = 'auth/invalid-email',
  InvalidPassword = 'auth/invalid-credential',
  TooManyRequsts = 'auth/too-many-requests',
}

export enum ImageType {
  Webp = 'image/webp',
  Jpeg = 'image/jpeg',
  Jpg = 'image/jpg',
  Png = 'image/png',
}

export interface IOption {
  name: string;
  value: string;
}

export enum ScrollDirection {
  Up = 'up',
  Down = 'down',
}

export enum Severity {
  Success = 'success',
  Info = 'info',
  Warning = 'warning',
  Error = 'error',
}

export interface ICurrentRecipe extends Partial<IRecipe> {}
