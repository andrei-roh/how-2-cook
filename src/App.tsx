import { Outlet } from 'react-router-dom';
import css from './App.module.sass';
import { Header, Notification } from './components';

const App = () => {
  return (
    <>
      <Header />
      <div className={css.app}>
        <Outlet />
      </div>
      <Notification />
    </>
  );
};

export default App;
