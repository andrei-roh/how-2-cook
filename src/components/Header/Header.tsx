import { useDispatch, useSelector } from 'react-redux';
import css from './Header.module.sass';
import { IState, IUser } from 'src/types';
import { Button } from '../Button/Button';
import { getAuth, signOut } from 'firebase/auth';
import { showNotification } from 'src/utils/showNotification';
import { NOTIFICATION, ROOT_ROUTE } from 'src/constants';
import { useState } from 'react';
import { Modal } from '../Modal/Modal';
import Skeleton from 'src/assets/fish-skeleton.svg';
import { useNavigate } from 'react-router-dom';
import { Dispatch } from '@reduxjs/toolkit';
import { setUser } from 'src/redux/actions';

export const Header = () => {
  const navigate = useNavigate();
  const dispatch: Dispatch = useDispatch();
  const user = useSelector((state: IState) => state.user);
  const [isSignOut, setIsSignOut] = useState(false);

  const handleShowModal = () => setIsSignOut(true);
  const handleCloseModal = () => setIsSignOut(false);

  const handleSignOut = async () => {
    const auth = getAuth();

    signOut(auth)
      .then(() => {
        dispatch(setUser({} as IUser));
        showNotification(NOTIFICATION(user.email).SIGN_OUT_SUCCESS, 6000);
        navigate(ROOT_ROUTE);
      })
      .catch(() => {
        showNotification(NOTIFICATION(user.email).SIGN_OUT_ERROR, 6000);
      });
    
    handleCloseModal();
  };

  return (
    <div className={css.headerWrapper}>
      <div className={css.userEmail}>{user.email || ''}</div>
      {user.email && <Button onClick={handleShowModal} className={css.headerModalButton}>
        <img className={css.headerLogo} src={Skeleton} alt='Sign Out Button' />
      </Button>}
      {isSignOut && (
        <Modal
          cancelButtonMessage={'Отмена'}
          submitButtonMessage={'Выход'}
          isLoading={false}
          handleClose={handleCloseModal}
          handleSubmit={handleSignOut}
          message='Вы уверены, что хотите выйти?'
          submitClassName={css.modalUserLogoutButton}
          cancelClassName={css.modalCancelButton}
        />
      )}
    </div>
  );
};
