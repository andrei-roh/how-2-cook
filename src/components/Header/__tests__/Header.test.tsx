import { describe, expect, it, vi } from 'vitest';
import { render } from '@testing-library/react';
import { Header } from '../Header';
import { Provider } from 'react-redux';
import { IAction, IUser } from 'src/types';
import { SET_USER } from 'src/redux/types';
import { configureStore } from '@reduxjs/toolkit';

const userMock = { email: 'Test@email.com' };
const navigateMock = vi.fn();
const stateMock = {
  user: userMock as IUser,
  loading: false,
  error: null,
};

const reducerMock = (( state = stateMock, action: IAction) => {
  if (action.type === SET_USER) {
    return {
      ...state,
      user: action.payload,
    };
  }

  return state;
})

const storeMock = configureStore({ reducer: reducerMock });

vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');

  return {
    ...actual,
    useNavigate: () => ({ navigate: navigateMock }),
  };
});

describe('Testing: Header', () => {
  it('should render', () => {
    const { container } = render(
      <Provider store={storeMock}>
        <Header />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
