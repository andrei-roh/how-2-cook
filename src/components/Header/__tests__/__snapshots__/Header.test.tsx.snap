// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing: Header > should render 1`] = `
<div>
  <div
    class="_headerWrapper_069298"
  >
    <div
      class="_userEmail_069298"
    >
      Test@email.com
    </div>
    <button
      class="_buttonWrapper_89d844 _headerModalButton_069298"
    >
      <img
        alt="Sign Out Button"
        class="_headerLogo_069298"
        src="/src/assets/fish-skeleton.svg"
      />
    </button>
  </div>
</div>
`;
